v2/internal/identity/credential_provider.go

```diff
@@ -0,0 +1,260 @@
+	return reflect.DeepEqual(c.secretData, other.secretData)
+}
+
+func NewDefaultCredential(tokenCred azcore.TokenCredential, namespace string, subscriptionID string) *Credential {
+	return &Credential{
+		TokenCredential: tokenCred,
+		SubscriptionID:  subscriptionID,
+		CredentialFrom:  types.NamespacedName{Namespace: namespace, Name: globalCredentialSecretName},
+	}
+}
+
+type CredentialProvider interface {
+	GetCredential(ctx context.Context, obj genruntime.MetaObject) (*Credential, error)
+}
+
+type credentialProvider struct {
+	globalCredential *Credential
+	kubeClient       kubeclient.Client
+}
+
+func NewCredentialProvider(
+	globalCredential *Credential,
+	kubeClient kubeclient.Client,
+) CredentialProvider {
+	return &credentialProvider{
+		kubeClient:       kubeClient,
+		globalCredential: globalCredential,
+	}
+}
+
+// GetCredential returns the credential used to manage the obj. It performs credential lookup according to the following
+// hierarchy:
+// 1. Per-Resource credential specified as an annotation (serviceoperator.azure.com/credential-from) on the obj
+// 2. Per-Namespace credential provided at the namespace scope ("aso-credential" secret in the obj's namespace)
+// 3. Global credential for the operator.
+//
+// If no matching credential can be found, an error is returned.
+func (c *credentialProvider) GetCredential(ctx context.Context, obj genruntime.MetaObject) (*Credential, error) {
+	cred, err := c.getCredentialFromAnnotation(ctx, obj, reconcilers.PerResourceSecretAnnotation)
+	if err != nil {
+		return nil, err
+	}
+	if cred != nil {
+		return cred, nil
+	}
+
+	// Namespaced secret
+	cred, err = c.getCredentialFromNamespaceSecret(ctx, obj.GetNamespace())
+	if err != nil {
+		return nil, err
+	}
+	if cred != nil {
+		return cred, nil
+	}
+
+	if c.globalCredential == nil {
+		return nil, errors.New("global credential not configured, you must use either namespaced or per-resource credentials")
+	}
+	// If not found, default is the global credential
+	return c.globalCredential, nil
+}
+
+func (c *credentialProvider) getCredentialFromNamespaceSecret(ctx context.Context, namespace string) (*Credential, error) {
+	secretNamespacedName := types.NamespacedName{Namespace: namespace, Name: NamespacedSecretName}
+	cred, err := c.getCredentialFromSecret(ctx, secretNamespacedName)
+	if err != nil {
+		var secretNotFound *core.SecretNotFound
+		if errors.As(err, &secretNotFound) {
+			return nil, nil // Not finding this secret is allowed, allow caller to proceed to higher scope secret
+		}
+		if apierrors.IsNotFound(err) {
+
+		}
+		return nil, err
+	}
+
+	return cred, nil
+}
+
+func (c *credentialProvider) getCredentialFromAnnotation(ctx context.Context, obj genruntime.MetaObject, annotation string) (*Credential, error) {
+	credentialFrom, ok := obj.GetAnnotations()[annotation]
+	if !ok {
+		return nil, nil
+	}
+
+	// Disallow credentials with `/` in their credentialFrom
+	if strings.Contains(credentialFrom, "/") {
```

Comment:
Are there any potential encodings we need to be aware of here?

E.g. we're blocking `/` but not `%2F`

Reply:
No, in fact this `if` isn't really required to be "safe" - it's just for a more clear user error. If the user passed us a secret with name: `my/secret`, we would lookup a secret with: `types.NamespacedName{Namespace: obj.Namespace, Name: "my/secret"}`. That lookup will fail because no such secret exists.